### ISSUE TITLE
Clients keep losing connection after upgrade to 2.9.0.1

### ISSUE THREAD

**Issue by christianhennen** [Icinga/icinga2#6466]
After upgrading from 2.8.4.1 to 2.9.0.1 all clients in our setup keep disconnecting ca. 2 times an hour and directly reconnect afterwards. The problem appeared right after the upgrade. Nodes that have not been upgraded yet do not show this behavior.

## Expected Behavior
Endpoints do only disconnect when there is a network or certificate issue or something similar.

## Current Behavior
Endpoints disconnect twice per hour and reconnect within 20 seconds.

## Steps to Reproduce (for bugs)
apt upgrade to 2.9.0.1

## Context
Before the update everything worked as expected. There seems to be a hostname problem for some endpoints, but it seems to be unrelated since every updated endpoint shows the behavior described above, also the correctly named ones.

## Your Environment
* Version used (`icinga2 --version`): 2.9.0.1
* Operating System and version: Ubuntu 16.04 LTS on master, Ubuntu 14.04 and 16.04 on nodes
* Enabled features (`icinga2 feature list`): api checker ido-mysql influxdb mainlog notification
* Icinga Web 2 version and modules (System - About): 2.6.0, Modules: monitoring
* Config validation (`icinga2 daemon -C`): 
```
[2018-07-20 12:48:05 +0200] information/cli: Icinga application loader (version: r2.9.0-1)
[2018-07-20 12:48:05 +0200] information/cli: Loading configuration file(s).
[2018-07-20 12:48:05 +0200] information/ConfigItem: Committing config item(s).
[2018-07-20 12:48:05 +0200] information/ApiListener: My API identity: icinga2.uni-trier.de
[2018-07-20 12:48:06 +0200] information/ConfigItem: Instantiated 2229 Services.
[2018-07-20 12:48:06 +0200] information/ConfigItem: Instantiated 3 ServiceGroups.
[2018-07-20 12:48:06 +0200] information/ConfigItem: Instantiated 1 ScheduledDowntime.
[2018-07-20 12:48:06 +0200] information/ConfigItem: Instantiated 1 InfluxdbWriter.
[2018-07-20 12:48:06 +0200] information/ConfigItem: Instantiated 37 HostGroups.
[2018-07-20 12:48:06 +0200] information/ConfigItem: Instantiated 1 FileLogger.
[2018-07-20 12:48:06 +0200] information/ConfigItem: Instantiated 1 NotificationComponent.
[2018-07-20 12:48:06 +0200] information/ConfigItem: Instantiated 2 NotificationCommands.
[2018-07-20 12:48:06 +0200] information/ConfigItem: Instantiated 2483 Notifications.
[2018-07-20 12:48:06 +0200] information/ConfigItem: Instantiated 1 IcingaApplication.
[2018-07-20 12:48:06 +0200] information/ConfigItem: Instantiated 261 Hosts.
[2018-07-20 12:48:06 +0200] information/ConfigItem: Instantiated 1 ApiListener.
[2018-07-20 12:48:06 +0200] information/ConfigItem: Instantiated 3 Downtimes.
[2018-07-20 12:48:06 +0200] information/ConfigItem: Instantiated 946 Dependencies.
[2018-07-20 12:48:06 +0200] information/ConfigItem: Instantiated 7 Comments.
[2018-07-20 12:48:06 +0200] information/ConfigItem: Instantiated 1 CheckerComponent.
[2018-07-20 12:48:06 +0200] information/ConfigItem: Instantiated 95 Zones.
[2018-07-20 12:48:06 +0200] information/ConfigItem: Instantiated 93 Endpoints.
[2018-07-20 12:48:06 +0200] information/ConfigItem: Instantiated 3 ApiUsers.
[2018-07-20 12:48:06 +0200] information/ConfigItem: Instantiated 5 UserGroups.
[2018-07-20 12:48:06 +0200] information/ConfigItem: Instantiated 1 IdoMysqlConnection.
[2018-07-20 12:48:06 +0200] information/ConfigItem: Instantiated 216 CheckCommands.
[2018-07-20 12:48:06 +0200] information/ConfigItem: Instantiated 3 TimePeriods.
[2018-07-20 12:48:06 +0200] information/ConfigItem: Instantiated 6 Users.
[2018-07-20 12:48:06 +0200] information/ScriptGlobal: Dumping variables to file '/var/cache/icinga2/icinga2.vars'
[2018-07-20 12:48:06 +0200] information/cli: Finished validating the configuration file(s).
```

Example of an Endpoint (every other endpoint is configured the same way):
```
Object 'ns02.uni-trier.de' of type 'Endpoint':
  % declared in '/etc/icinga2/zones.d/master/ansible_hosts/ns02.uni-trier.de.conf', lines 37:1-37:35
  * __name = "ns02.uni-trier.de"
  * host = "ns02.uni-trier.de"
    % = modified in '/etc/icinga2/zones.d/master/ansible_hosts/ns02.uni-trier.de.conf', lines 38:3-38:28
  * log_duration = 86400
  * name = "ns02.uni-trier.de"
  * package = "_etc"
  * port = "5665"
  * source_location
    * first_column = 1
    * first_line = 37
    * last_column = 35
    * last_line = 37
    * path = "/etc/icinga2/zones.d/master/ansible_hosts/ns02.uni-trier.de.conf"
  * templates = [ "ns02.uni-trier.de" ]
    % = modified in '/etc/icinga2/zones.d/master/ansible_hosts/ns02.uni-trier.de.conf', lines 37:1-37:35
  * type = "Endpoint"
  * zone = "master"
```

Example of a Zone (every other zone is configured the same way):
```
Object 'ns02.uni-trier.de' of type 'Zone':
  % declared in '/etc/icinga2/zones.d/master/ansible_hosts/ns02.uni-trier.de.conf', lines 32:1-32:31
  * __name = "ns02.uni-trier.de"
  * endpoints = [ "ns02.uni-trier.de" ]
    % = modified in '/etc/icinga2/zones.d/master/ansible_hosts/ns02.uni-trier.de.conf', lines 33:3-33:37
  * global = false
  * name = "ns02.uni-trier.de"
  * package = "_etc"
  * parent = "master"
    % = modified in '/etc/icinga2/zones.d/master/ansible_hosts/ns02.uni-trier.de.conf', lines 34:3-34:19
  * source_location
    * first_column = 1
    * first_line = 32
    * last_column = 31
    * last_line = 32
    * path = "/etc/icinga2/zones.d/master/ansible_hosts/ns02.uni-trier.de.conf"
  * templates = [ "ns02.uni-trier.de" ]
    % = modified in '/etc/icinga2/zones.d/master/ansible_hosts/ns02.uni-trier.de.conf', lines 32:1-32:31
  * type = "Zone"
  * zone = "master"
```

**Comment by christianhennen** [comment]
Sry, forgot to paste it:

```
[2018-07-20 13:15:37 +0200] information/JsonRpcConnection: No messages for identity 'tares.uni-trier.de' have been received in the last 60 seconds.
[2018-07-20 13:15:37 +0200] warning/JsonRpcConnection: API client disconnected for identity 'tares.uni-trier.de'
[2018-07-20 13:15:37 +0200] warning/JsonRpcConnection: API client disconnected for identity 'tares.uni-trier.de'
[2018-07-20 13:15:37 +0200] warning/ApiListener: Removing API client for endpoint 'tares.uni-trier.de'. 0 API clients left.
[2018-07-20 13:15:37 +0200] warning/ApiListener: Removing API client for endpoint 'tares.uni-trier.de'. 0 API clients left.
[2018-07-20 13:15:57 +0200] information/ApiListener: Reconnecting to endpoint 'tares.uni-trier.de' via host 'tares.uni-trier.de' and port '5665'
[2018-07-20 13:15:57 +0200] information/ApiListener: New client connection for identity 'tares.uni-trier.de' to [136.199.199.101]:5665
[2018-07-20 13:15:57 +0200] information/ApiListener: Finished reconnecting to endpoint 'tares.uni-trier.de' via host 'tares.uni-trier.de' and port '5665'
[2018-07-20 13:15:57 +0200] information/ApiListener: Sending config updates for endpoint 'tares.uni-trier.de' in zone 'tares.uni-trier.de'.
[2018-07-20 13:15:57 +0200] information/JsonRpcConnection: Received certificate request for CN 'tares.uni-trier.de' signed by our CA.
[2018-07-20 13:15:57 +0200] information/JsonRpcConnection: The certificate for CN 'tares.uni-trier.de' is valid and uptodate. Skipping automated renewal.
[2018-07-20 13:15:57 +0200] information/ApiListener: Syncing configuration files for global zone 'global-templates' to endpoint 'tares.uni-trier.de'.
[2018-07-20 13:15:57 +0200] information/ApiListener: Finished sending config file updates for endpoint 'tares.uni-trier.de' in zone 'tares.uni-trier.de'.
[2018-07-20 13:15:57 +0200] information/ApiListener: Syncing runtime objects to endpoint 'tares.uni-trier.de'.
[2018-07-20 13:15:57 +0200] information/ApiListener: Finished syncing runtime objects to endpoint 'tares.uni-trier.de'.
[2018-07-20 13:15:57 +0200] information/ApiListener: Finished sending runtime config updates for endpoint 'tares.uni-trier.de' in zone 'tares.uni-trier.de'.
[2018-07-20 13:15:57 +0200] information/ApiListener: Sending replay log for endpoint 'tares.uni-trier.de' in zone 'tares.uni-trier.de'.
[2018-07-20 13:15:57 +0200] information/ApiListener: Finished sending replay log for endpoint 'tares.uni-trier.de' in zone 'tares.uni-trier.de'.
[2018-07-20 13:15:57 +0200] information/ApiListener: Finished syncing endpoint 'tares.uni-trier.de' in zone 'tares.uni-trier.de'.
```
Currently we are not experiencing any network issues. The VM has plenty of resources and the connection worked fine since installation last year. The problem seems to be directly linked to the update. A colleague of mine updated some servers the day before yesterday and they started acting up. As I updated some others yesterday, they too became affected. Servers that have not yet been updated do not show this behavior.

**Comment by anthonymf** [comment]
After upgrading I saw clients disconnecting after 60s if no service check reported back. Adding a service check with a check interval of 30s resolved it for me.

This is anecdotal but may help.

**Comment by christianhennen** [comment]
Hm... I already suspected that this might not be a bug but rather some normal change that has been introduced with this version. 
I assumed the cluster check would be performed by the master asking the endpoints if they're available. But looking at the log output again and taking your info into consideration, it seems that the clients need to regularly check in. Maybe the amount of time the master waits for a lifesign of the client has been reduced?

**Comment by christianhennen** [comment]
Strange... I just wanted to check if setting the interval of a service check to 30s would eliminate the problem. But beforehand I re-checked the logs and saw that the problem disappeared since around noon. The only thing I did then however was to set load_percpu to true, disable notifications for cluster-zone and reload the configs. Since the problems arised, I already restarted the Icinga host...

**Comment by Crunsher** [comment]
I wasn't able to understand from your last post whether you are still experiencing this issue. If not, what exactly did you change?

**Comment by christianhennen** [comment]
Right now, we are not experiencing any issues. The log messages about disconnecting endpoints have ended yesterday around noon. At the same time, I set "enable_notifications" of the check command "cluster-zone" to false and "load_percpu" of the check command load to true (see below). 
After my last post, I re-enabled notifications for "cluster-zone" but the issue didn't occur since then.

```
apply Service "load" {
  import "generic-service"

  check_command = "load"

  /* Used by the ScheduledDowntime apply rule in `downtimes.conf`. */
  vars.backup_downtime = "02:00-03:00"
  vars.load_percpu = true

  assign where host.name == NodeName
}
```
```
apply Service "cluster-health" {
  check_command = "cluster-zone"

  display_name = "cluster-health-" + host.name

  /* This follows the convention that the client zone name is the FQDN which is the same as the host object name. */
  vars.cluster_zone = host.name

  enable_notifications = false

  assign where host.vars.client_endpoint
}
```
What's weird about all this is the fact, that before yesterday we didn't change anything in the configuration. We installed the cluster (1 master, 93 endpoints and some other hosts) last year, with the clients using the API/Agent instead of NRPE like our old monitoring setup. The config is fairly standard, with checks being defined as a global template and applied if specific variables match.
The problems appeared just as I updated the master and some clients and didn't disappear after I reloaded the config or restarted the master. Yesterday however, after the changes mentioned above, the errors stopped.
